samples/order/processor/processor.go

```diff
@@ -0,0 +1,251 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A processor processes images in orders. It receives requests by
+// subscribing to the requests topic, and writes responses to the response
+// topic.
+//
+// At present, it just encodes input images into PNG format.
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"image"
+	_ "image/jpeg"
+	"image/png"
+	"log"
+	"time"
+
+	"gocloud.dev/blob"
+	_ "gocloud.dev/blob/fileblob"
+	"gocloud.dev/docstore"
+	_ "gocloud.dev/docstore/memdocstore"
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	_ "gocloud.dev/pubsub/mempubsub"
+	"gocloud.dev/samples/order/internal/common"
+)
+
+var (
+	requestSubURL    = flag.String("request-sub", "", "URL for request pub/sub subscription")
+	responseTopicURL = flag.String("response-topic", "", "URL for response pub/sub topic")
+	bucketURL        = flag.String("bucket", "", "URL for image bucket")
+	collectionURL    = flag.String("collection", "", "URL for order collection")
+)
+
+func main() {
+	flag.Parse()
+	if *requestSubURL == "" {
+		log.Fatal("missing -request-sub")
+	}
+	if *responseTopicURL == "" {
+		log.Fatal("missing -response-topic")
+	}
+	if *bucketURL == "" {
+		log.Fatal("missing -bucket")
+	}
+	if *collectionURL == "" {
+		log.Fatal("missing -collection")
+	}
+	ctx := context.Background()
+	p, err := newProcessor(ctx, *requestSubURL, *responseTopicURL, *bucketURL, *collectionURL)
+	if err != nil {
+		log.Fatal(err)
+	}
+	if err := p.run(ctx); err != nil {
+		log.Fatal(err)
+	}
+}
+
+type processor struct {
+	requestSub    *pubsub.Subscription
+	responseTopic *pubsub.Topic
+	bucket        *blob.Bucket
+	coll          *docstore.Collection
+}
+
+func newProcessor(ctx context.Context, subURL, topicURL, bucketURL, collURL string) (*processor, error) {
+	reqSub, err := pubsub.OpenSubscription(ctx, subURL)
+	if err != nil {
+		return nil, err
+	}
+	resTopic, err := pubsub.OpenTopic(ctx, topicURL)
+	if err != nil {
+		return nil, err
+	}
+	bucket, err := blob.OpenBucket(ctx, bucketURL)
+	if err != nil {
+		return nil, err
+	}
+	coll, err := docstore.OpenCollection(ctx, collURL)
+	if err != nil {
+		return nil, err
+	}
+	return &processor{
+		requestSub:    reqSub,
+		responseTopic: resTopic,
+		bucket:        bucket,
+		coll:          coll,
+	}, nil
+}
+
+func (p *processor) run(ctx context.Context) error {
+	for {
+		if err := p.handleRequest(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *processor) handleRequest(ctx context.Context) error {
+	var req common.OrderRequest
+	msg, err := receiveJSON(ctx, p.requestSub, &req)
+	// Ack the message because we handled it, even on error.
+	defer func() {
+		if msg != nil {
+			msg.Ack()
+		}
+	}()
+	if err != nil {
```

Comment:
If `err != nil`, is `msg` ever non-nil? This seems like it would be clearer like:

```
msg, err := ...
if err != nil {
  return err
}
defer func() {
  msg.Ack()
}
```

Reply:
I have to consider the case where I received a message, but couldn't JSON-unmarshal it. `receiveJSON` returns a non-nil message in that case, as documented.